Source/Core/Core/Debugger/BranchWatch.cpp

```diff
@@ -0,0 +1,312 @@
+// Copyright 2024 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Core/Debugger/BranchWatch.h"
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdio>
+
+#include <fmt/format.h>
+
+#include "Common/Assert.h"
+#include "Common/BitField.h"
+#include "Common/CommonTypes.h"
+#include "Core/Core.h"
+#include "Core/PowerPC/Gekko.h"
+#include "Core/PowerPC/MMU.h"
+
+namespace Core
+{
+void BranchWatch::Clear(const CPUThreadGuard&)
+{
+  m_selection.clear();
+  m_collection_vt.clear();
+  m_collection_vf.clear();
+  m_collection_pt.clear();
+  m_collection_pf.clear();
+  m_recording_phase = Phase::Blacklist;
+  m_blacklist_size = 0;
+}
+
+union USnapshotMetadata
+{
+  using Inspection = BranchWatch::SelectionInspection;
+  using StorageType = unsigned long long;
+
+  static_assert(Inspection::EndOfEnumeration == Inspection{(1u << 3) + 1});
+
+  StorageType hex;
+
+  BitField<0, 1, bool, StorageType> is_virtual;
+  BitField<1, 1, bool, StorageType> condition;
+  BitField<2, 1, bool, StorageType> is_selected;
+  BitField<3, 4, Inspection, StorageType> inspection;
+
+  USnapshotMetadata() : hex(0) {}
+  explicit USnapshotMetadata(bool is_virtual_, bool condition_, bool is_selected_,
+                             Inspection inspection_)
+      : USnapshotMetadata()
+  {
+    is_virtual = is_virtual_;
+    condition = condition_;
+    is_selected = is_selected_;
+    inspection = inspection_;
+  }
+};
+
+void BranchWatch::Save(const CPUThreadGuard& guard, std::FILE* file) const
+{
+  if (!CanSave())
+  {
+    ASSERT_MSG(CORE, false, "BranchWatch can not be saved.");
+    return;
+  }
+  if (file == nullptr)
+    return;
+
+  const auto save_collection = [&](Collection BranchWatch::*collection, bool is_virtual,
+                                   bool condition) {
```

Comment:
Likewise here, why can you not just pass a `const Collection&` and get rid of the somewhat confusing pointer-to-member? There's a few more instances of this pattern in this file, too.

Reply:
Sometimes I get a little ahead of myself finding uses for lesser-used sects of the C++ language. You are right, I can just use Collection references in a few places.